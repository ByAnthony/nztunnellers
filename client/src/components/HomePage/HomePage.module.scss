@use 'src/index.scss';

.homepage {
  margin: 0 auto;
}

.homepage-container {
  margin: 0 auto;
  margin-top: index.$xl-margin;
  width: index.$inner-width;

  h2 {
    padding-bottom: index.$base-margin;
    font-size: 5rem;
    font-weight: 200;
    text-align: center;
    line-height: normal;

    span {
      display: block;
      padding-top: 8px;
      font-weight: 500;

      span {
        padding-right: index.$xs-margin;
        font-size: 2rem;
        font-weight: 200;
        line-height: 0;
      }
    }
  }
}

.chapter-cards-wrapper {
  padding: index.$l-margin 0 index.$base-margin 0;
  height: 325px;
  overflow: hidden;
  border-radius: index.$border-radius;
  background-color: index.$primary-medium-color;
}

.chapter-cards {
  margin: 0 auto;
  padding: 0 index.$l-margin;
  display: flex;
  height: 390px;
  flex-direction: row;
  flex-wrap: nowrap;
  gap: index.$m-margin;
  overflow-x: auto;
}

.chapter-card {
  height: 300px;
  width: 350px;
  border: 1px solid rgb(64, 66, 67);
  border-radius: index.$border-radius;
  text-align: center;
  transition: border-bottom 0.9s ease-in-out, border-bottom 0.9s ease-in-out,
    border 0.9s ease-in-out, border 0.9s ease-in-out,
    background-color 0.9s ease-in-out, background-color 0.9s ease-in-out;

  &:hover {
    border: 1px solid index.$secondary-color;
  }
}

.chapter-card-dimmer {
  height: 300px;
  width: 350px;
  border-radius: 6.5px;
  background-color: rgba(44, 46, 47, 0.9);
  transition: border-bottom 0.9s ease-in-out, border-bottom 0.9s ease-in-out,
    border 0.9s ease-in-out, border 0.9s ease-in-out,
    background-color 0.9s ease-in-out, background-color 0.9s ease-in-out;

  &:hover {
    background-color: rgba(29, 31, 32, 0.85);
  }
}

.chapter-card-content {
  display: flex;
  height: 100%;
  width: 100%;
  justify-content: center;
  align-items: center;

  .title-line-1 {
    font-size: 1.9rem;
    font-weight: 200;
    color: index.$secondary-color;
  }

  .title-line-2 {
    display: block;
    font-size: 1.9rem;
    font-weight: 500;
    color: index.$secondary-color;
  }

  .title-line-3 {
    display: block;
    padding-top: index.$m-margin;
    font-size: 1.1rem;
    font-weight: 500;
    color: index.$secondary-color;
  }
}
