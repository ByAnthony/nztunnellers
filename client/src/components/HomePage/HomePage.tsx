import { useEffect } from 'react';
import { useGetAllHistoryArticleLinkQuery } from '../../redux/slices/historySlice';
import { Menu } from '../Menu/Menu';

import STYLES from './HomePage.module.scss';

export function HomePage() {
  useEffect(() => {
    window.addEventListener('scroll', () => {
      const scrollPosition = window.scrollY;
      const translateValue = `translateX(${-scrollPosition}px)`;
      document.getElementById('chapterCards')!.style.transform = translateValue;
    });
  });

  const {
    data, error, isLoading, isSuccess,
  } = useGetAllHistoryArticleLinkQuery();

  if (data) {
    return (
      <div className={STYLES.homepage}>
        {error && (
        <div>
          <p>An error occured</p>
        </div>
        )}
        <Menu />
        { isLoading }
        { isSuccess && (
          <div className={STYLES['homepage-container']}>
            <h2>History of the Company</h2>
            <div className={STYLES['chapter-cards']} id="chapterCards">
              {data.map((article) => {
                const divStyle = {
                  backgroundImage: `url(../images/history/${article.image})`,
                  backgroundSize: 'cover',
                };

                return (
                  <div className={STYLES['chapter-card']} key={data.indexOf(article)} style={divStyle}>
                    <a
                      href={`/history/${article.url}`}
                      className={STYLES['link-button']}
                      aria-label={`Go to Chapter ${article.chapter}: ${article.title.replace('\\', ' ')}`}
                    >
                      <div>
                        <p>{article.chapter}</p>
                        <span>{article.title.replace('\\', ' ')}</span>
                      </div>
                    </a>
                  </div>
                );
              })}
            </div>
          </div>
        )}
        {/* <Footer /> */}
      </div>
    );
  }
  return null;
}
