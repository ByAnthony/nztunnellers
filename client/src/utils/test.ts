import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ReactElement } from 'react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';

export const findElementWithText = (text: string) => {
  const elements = screen.getAllByText((content) => content.includes(text));

  if (elements.length === 0) {
    throw new Error(`Unable to find an element with the text: ${text}`);
  }

  return elements[0];
};

export const renderWithMemoryRouter = (ui: ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);
  const user = userEvent.setup();

  return {
    user,
    ...render(ui, { wrapper: MemoryRouter }),
  };
};

const renderWithRouter = (ui: ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);
  const user = userEvent.setup();

  return {
    user,
    ...render(ui, { wrapper: BrowserRouter }),
  };
};

export * from '@testing-library/react';
export { renderWithRouter as render };
