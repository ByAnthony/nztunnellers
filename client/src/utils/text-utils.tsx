export const formatText = (text: string) => {
  const paragraphs = text.split('\\n\\n');
  let segmentKey: number = 0;
  return paragraphs.map((paragraph) => {
    const segments = paragraph
      .replace(/--/g, '\u00A0')
      .split(/(\*.+?\*)|(\[.+?\))/g);
    const formattedSegments = segments.map((segment) => {
      segmentKey += 1;
      if (segment && segment.startsWith('*') && segment.endsWith('*')) {
        const italicText = segment.slice(1, -1);
        return <em key={segmentKey}>{italicText}</em>;
      }
      if (segment && segment.startsWith('[') && segment.endsWith(')')) {
        const linkText = segment.slice(1, -1);
        const [label, url] = linkText.split('](');
        const number = label.slice(0, -1);
        if (url.includes('footnote')) {
          return (
            <a key={segmentKey} href={url} id={`reference_${number}`}>
              [
              {label}
              ]
            </a>
          );
        }
        return (
          <a key={segmentKey} href={url} id={`footnote_${number}`}>
            {label}
          </a>
        );
      }
      return <span key={segmentKey}>{segment}</span>;
    });
    return <p key={paragraphs.indexOf(paragraph)}>{formattedSegments}</p>;
  });
};
