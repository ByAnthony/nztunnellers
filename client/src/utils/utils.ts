export const displayBirthDeathDates = (
  birth: string | null,
  death: string | null,
) => {
  if (birth !== null && death !== null) {
    return `${birth} - ${death}`;
  }
  if (birth !== null && death === null) {
    return `${birth} - †?`;
  }
  if (birth === null && death !== null) {
    return `? - ${death}`;
  }
  return '? - †?';
};
