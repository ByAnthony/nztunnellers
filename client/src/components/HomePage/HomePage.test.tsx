import '@testing-library/jest-dom';
import { HomePage } from './HomePage';
import { renderWithMemoryRouter } from '../../utils/test-utils';

const component = <HomePage />;

test('renders the component correctly', () => {
  const { asFragment } = renderWithMemoryRouter(component);

  expect(asFragment()).toMatchSnapshot();
});
