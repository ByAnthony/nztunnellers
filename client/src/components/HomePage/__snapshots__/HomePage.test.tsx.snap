// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component correctly 1`] = `
<DocumentFragment>
  <div
    class="homepage"
  >
    <div
      class="menu"
    >
      <a
        aria-label="Go to the Homepage"
        class="logo"
        href="/"
      >
        <img
          alt=""
          src="/nzt_logo.png"
        />
      </a>
      <div
        class="search-form"
      >
        <form>
          <input
            id="search"
            name="search"
            placeholder="Search for a tunneller"
            type="text"
          />
        </form>
        <button
          class="search-form-button"
          type="submit"
        >
          <img
            alt=""
            height="18"
            src="/searching_hover.png"
          />
        </button>
      </div>
    </div>
    <div
      class="main"
    >
      <div>
        History
      </div>
      <div>
        Tunnellers
      </div>
    </div>
  </div>
</DocumentFragment>
`;
